File: 02_Libraries/ruby/kirin_dns.rb
# kirin_dns.rb — KirinDNS Resolution Protocol (ADRP) v2.0 Ruby Client
#
# SRV records for service port discovery (_kirinnet-http._tcp, etc.)
# TXT records for identity metadata (id=;key=;nick=;ipfs=)
# Legacy ADRP JSON TXT fallback preserved for backward compatibility.
#
# Pure Ruby — only stdlib dependencies (resolv, json).
#
# Usage:
#   require_relative 'kirin_dns'
#   srv = KirinDNS.resolve_service('alice.kirinnet.org', 'ws')
#   puts "WS: #{srv[:target]}:#{srv[:port]}"
#   id = KirinDNS.resolve_identity('alice.kirinnet.org')
#   puts "ID: #{id[:id]}"

require 'resolv'
require 'json'
require 'set'
require 'digest'
require 'base64'

module KirinDNS
  # SRV service names (spec Section 2.2)
  SRV_SERVICES = {
    http:  '_kirinnet-http._tcp',
    https: '_kirinnet-https._tcp',
    ws:    '_kirinnet-ws._tcp'
  }.freeze

  # Fallback ports (spec Section 3.3.1, Step 4)
  FALLBACK = { http: 80, https: 443, ws: 80, wss: 443 }.freeze

  # Legacy ADRP JSON recognized keys
  RECOGNIZED = %w[http https ws wss].to_set.freeze

  # ---------------------------------------------------------------------------
  # Service Resolution (SRV)
  # ---------------------------------------------------------------------------

  # Resolve a single service port via SRV.
  #
  # @param domain  [String] e.g., 'alice.kirinnet.org'
  # @param service [String] 'http', 'https', or 'ws'
  # @return [Hash{Symbol=>String,Integer}, nil]
  #   { target: 'alice.kirinnet.org', port: 8082 } or nil if no SRV record.
  def self.resolve_service(domain, service)
    srv_name = SRV_SERVICES[service.to_sym]
    raise ArgumentError, "Unknown service: #{service}. Recognized: http, https, ws" unless srv_name

    full_name = "#{srv_name}.#{domain}"

    begin
      resolver = Resolv::DNS.new
      records = resolver.getresources(full_name, Resolv::DNS::Resource::IN::SRV)
    rescue Resolv::ResolvError
      return nil
    end

    return nil if records.nil? || records.empty?

    # RFC 2782: lowest priority, then highest weight
    best = records.min_by { |r| [r.priority, -r.weight] }
    { target: best.target.to_s.sub(/\.\z/, ''), port: best.port }
  end

  # Resolve all SRV services for a domain.
  #
  # @param domain [String]
  # @return [Hash]
  #   { http: {target:, port:}|nil, https: ..., ws: ... }
  def self.resolve_all_services(domain)
    SRV_SERVICES.each_key.each_with_object({}) do |svc, hash|
      hash[svc] = resolve_service(domain, svc.to_s)
    end
  end

  # ---------------------------------------------------------------------------
  # Legacy Wrapper (all-in-one resolution)
  # ---------------------------------------------------------------------------

  # Full resolution: SRV + TXT identity.
  #
  # @param domain [String]
  # @return [Hash]
  #   { domain:, ws:, http:, https:, identity: }
  def self.resolve_kirin_dns(domain)
    {
      domain: domain,
      ws: resolve_service(domain, 'ws') || { target: domain, port: FALLBACK[:ws] },
      http: resolve_service(domain, 'http'),
      https: resolve_service(domain, 'https'),
      identity: resolve_identity(domain)
    }
  end

  # ---------------------------------------------------------------------------
  # Identity Resolution (TXT)
  # ---------------------------------------------------------------------------

  # Parse a semicolon-separated key=value TXT string into an identity hash.
  #
  # Format: id=<uuid>;key=<hex>;nick=<name>;ipfs=<bool>
  # (spec Section 3.2)
  #
  # @param txt [String] Raw TXT record value.
  # @return [Hash, nil] Parsed identity, or nil if not a valid identity record.
  def self.parse_identity_txt(txt)
    return nil if txt.nil? || txt.strip.empty?
    return nil unless txt.start_with?('id=')

    result = {}
    txt.split(';').each do |pair|
      eq = pair.index('=')
      next unless eq
      key = pair[0...eq].strip
      val = pair[(eq + 1)..].strip
      result[key] = val
    end

    # Both id and key are required
    return nil unless result['id'] && result['key']

    # Parse ipfs boolean if present
    if result.key?('ipfs')
      result['ipfs'] = result['ipfs'].downcase == 'true'
    end

    result.transform_keys(&:to_sym)
  end

  # Resolve identity metadata from TXT records.
  #
  # @param domain [String]
  # @return [Hash, nil]
  #   { id:, key:, nick?:, ipfs?: } or nil if no identity TXT found.
  def self.resolve_identity(domain)
    begin
      resolver = Resolv::DNS.new
      records = resolver.getresources(domain, Resolv::DNS::Resource::IN::TXT)
    rescue Resolv::ResolvError
      return nil
    end

    records.each do |record|
      txt = record.strings.join('')
      identity = parse_identity_txt(txt)
      return identity if identity
    end

    nil
  end

  # ---------------------------------------------------------------------------
  # Legacy ADRP API (backward compatibility)
  # ---------------------------------------------------------------------------

  # Resolve KirinDNS ports for a domain using legacy ADRP JSON TXT.
  #
  # @param domain [String]
  # @return [Hash] with keys :http, :https, :ws, :wss
  def self.resolve(domain)
    ports = FALLBACK.dup

    begin
      resolver = Resolv::DNS.new
      records = resolver.getresources(domain, Resolv::DNS::Resource::IN::TXT)
    rescue Resolv::ResolvError
      return ports
    end

    records.each do |record|
      txt = record.strings.join('')
      parsed = parse_txt(txt)
      next unless parsed

      ports.merge!(parsed)
      return ports
    end

    ports
  end

  # Resolve using a custom DNS server (legacy API).
  def self.resolve_with_server(domain, dns_server)
    ports = FALLBACK.dup

    begin
      resolver = Resolv::DNS.new(nameserver: [dns_server])
      records = resolver.getresources(domain, Resolv::DNS::Resource::IN::TXT)
    rescue Resolv::ResolvError
      return ports
    end

    records.each do |record|
      txt = record.strings.join('')
      parsed = parse_txt(txt)
      next unless parsed

      ports.merge!(parsed)
      return ports
    end

    ports
  end

  # Parse a TXT record string as legacy ADRP JSON.
  def self.parse_txt(txt)
    return nil if txt.nil? || txt.strip.empty?

    begin
      data = JSON.parse(txt)
    rescue JSON::ParserError
      return nil
    end

    return nil unless data.is_a?(Hash)
    return nil if data.empty?

    result = {}
    RECOGNIZED.each do |key|
      val = data[key]
      next if val.nil?
      val = val.to_i if val.is_a?(String) && val.match?(/\A\d+\z/)
      return nil unless val.is_a?(Integer)
      return nil if val < 1 || val > 65535
      result[key.to_sym] = val
    end

    return nil if result.empty?
    result
  end

  # ---------------------------------------------------------------------------
  # did:dns three-record identity model (spec §3.2.1 / did-dns-protocol §2)
  # ---------------------------------------------------------------------------
  #
  # Tamper-evident fingerprint chain: fp == Base64URL(SHA-256(pk)[0:12]).
  # Uses Ruby stdlib Digest::SHA256 + Base64.urlsafe_encode/decode.

  DID_DNS_PREFIX       = 'did:dns:'.freeze
  DID_DNS_DECL_PREFIX  = 'did:dns:v='.freeze
  DID_DNS_PK_PREFIX    = 'did:dns:pk;'.freeze
  DID_DNS_BLACK_PREFIX = 'did:dns:black;'.freeze
  DID_DNS_KTY_ED25519  = 'ed25519'.freeze
  DID_DNS_FINGERPRINT_BYTES = 12   # -> 16 base64url chars

  # Base64URL decode (no padding). Returns the raw binary string, or nil.
  def self.base64url_decode(str)
    return nil if str.nil? || str.empty?
    pad = str.length % 4
    str = str + ('=' * (4 - pad)) if pad > 0
    Base64.urlsafe_decode64(str)
  rescue ArgumentError
    nil
  end

  # Base64URL encode (no padding).
  def self.base64url_encode(binary)
    Base64.urlsafe_encode64(binary, padding: false)
  end

  # Create a new DidDnsIdentity hash with defaults.
  def self.new_did_dns_identity
    {
      version: 1,
      fingerprint: '',
      nickname: '',       # Base64URL(UTF-8)
      gender: '',         # M/F/O/X
      issued_at: 0,
      expires_at: 0,
      key_type: DID_DNS_KTY_ED25519,
      public_key_b64url: '',
      blacklist: []
    }
  end

  # Recompute fp = Base64URL(SHA-256(pk)[0:12]). Empty string on malformed pk.
  def self.compute_fingerprint(id)
    pk = base64url_decode(id[:public_key_b64url])
    return '' if pk.nil? || pk.empty?
    digest = Digest::SHA256.digest(pk)
    base64url_encode(digest[0, DID_DNS_FINGERPRINT_BYTES])
  end

  # True iff declared fp matches recomputed fp over the pk bytes.
  def self.fingerprint_chain_ok?(id)
    !id[:fingerprint].empty? && id[:fingerprint] == compute_fingerprint(id)
  end

  def self.revoked?(id)
    id[:blacklist].include?(id[:fingerprint])
  end

  def self.expired?(id, now)
    id[:expires_at] != 0 && now >= id[:expires_at]
  end

  # Composite policy: v1 + ed25519 + chain + not revoked + not expired.
  def self.did_dns_valid?(id, now)
    id[:version] == 1 &&
      id[:key_type] == DID_DNS_KTY_ED25519 &&
      fingerprint_chain_ok?(id) &&
      !revoked?(id) &&
      !expired?(id, now)
  end

  # Parse a `k=v;k=v` segment (after the did:dns: prefix) into a hash.
  def self.parse_did_dns_kv(segment)
    out = {}
    segment.split(';').each do |pair|
      eq = pair.index('=')
      next unless eq
      k = pair[0...eq].strip
      v = pair[(eq + 1)..].strip
      out[k] = v
    end
    out
  end

  # Classify TXT records by did:dns: sub-type; assemble an identity.
  # Returns nil if no did:dns records / declaration+pk missing.
  def self.parse_did_dns_identity(txt_records)
    decl_raw = pk_raw = black_raw = nil
    txt_records.each do |raw|
      s = raw.to_s.strip
      if decl_raw.nil? && s.start_with?(DID_DNS_DECL_PREFIX)
        decl_raw = s
      elsif pk_raw.nil? && s.start_with?(DID_DNS_PK_PREFIX)
        pk_raw = s
      elsif black_raw.nil? && s.start_with?(DID_DNS_BLACK_PREFIX)
        black_raw = s
      end
    end
    return nil if decl_raw.nil? || pk_raw.nil?

    decl = parse_did_dns_kv(decl_raw[DID_DNS_PREFIX.length..])
    pk   = parse_did_dns_kv(pk_raw[DID_DNS_PREFIX.length..])

    id = new_did_dns_identity
    id[:version]           = decl.key?('v') ? decl['v'].to_i : 1
    id[:fingerprint]       = decl['fp'] || ''
    id[:nickname]          = decl['n'] || ''
    id[:gender]            = decl['g'] || ''
    id[:issued_at]         = decl.key?('iat') ? decl['iat'].to_i : 0
    id[:expires_at]        = decl.key?('exp') ? decl['exp'].to_i : 0
    id[:key_type]          = pk['kty'] || DID_DNS_KTY_ED25519
    id[:public_key_b64url] = pk['pk'] || ''

    unless black_raw.nil?
      bkv = parse_did_dns_kv(black_raw[DID_DNS_PREFIX.length..])
      fp_field = bkv['fp'] || ''
      id[:blacklist] = fp_field.split(',').reject(&:empty?)
    end
    id
  end

  # ---- self-test ----------------------------------------------------------
  if __FILE__ == $PROGRAM_NAME
    # Legacy parse tests
    p = parse_txt('{"http":8080,"https":8443}')
    raise 'valid parse failed' unless p
    raise 'http' unless p[:http] == 8080
    raise 'https' unless p[:https] == 8443

    raise 'empty should be nil' if parse_txt('{}')
    raise 'port zero should be nil' if parse_txt('{"http":0}')
    raise 'not json should be nil' if parse_txt('not json')

    # Resolution test
    ports = resolve('nonexistent.invalid')
    raise 'fallback http' unless ports[:http] == 80
    raise 'fallback https' unless ports[:https] == 443

    # ---- v2 identity parser tests ----

    parsed = parse_identity_txt(
      'id=550e8400-e29b-41d4-a716-446655440000;key=04abc;nick=Alice;ipfs=false'
    )
    raise 'valid identity failed' unless parsed
    raise 'id' unless parsed[:id] == '550e8400-e29b-41d4-a716-446655440000'
    raise 'key' unless parsed[:key] == '04abc'
    raise 'nick' unless parsed[:nick] == 'Alice'
    raise 'ipfs false' unless parsed[:ipfs] == false

    # Minimal identity
    minimal = parse_identity_txt('id=test-id;key=0x00')
    raise 'minimal id' unless minimal[:id] == 'test-id'
    raise 'minimal key' unless minimal[:key] == '0x00'
    raise 'no nick' if minimal.key?(:nick)

    # Invalid identities
    raise 'not identity' if parse_identity_txt('not an identity')
    raise 'spf record' if parse_identity_txt('v=spf1 include:_spf.example.com')
    raise 'empty' if parse_identity_txt('')
    raise 'nil' if parse_identity_txt(nil)
    raise 'no key' if parse_identity_txt('id=foo;nick=Bar')

    # ---- v2 SRV tests ----

    srv = resolve_service('nonexistent.invalid', 'ws')
    raise 'nil srv for nonexistent' unless srv.nil?

    begin
      resolve_service('example.com', 'bogus')
      raise 'should have raised'
    rescue ArgumentError
      # expected
    end

    all_srv = resolve_all_services('nonexistent.invalid')
    raise 'all nil' unless all_srv.values.all?(&:nil?)

    identity = resolve_identity('nonexistent.invalid')
    raise 'nil identity for nonexistent' unless identity.nil?

    puts 'KirinDNS Ruby self-test: PASSED'

    # ---- did:dns three-record identity model (C-1 baseline) ----
    # Deterministic 32-byte key = bytes 0..31 (matches the golden vector).
    pk_bytes = (0..31).to_a.pack('C*')
    pk_b64 = base64url_encode(pk_bytes)
    dg = Digest::SHA256.digest(pk_bytes)
    fp_calc = base64url_encode(dg[0, DID_DNS_FINGERPRINT_BYTES])
    now = 1_700_000_000

    recs = [
      'v=spf1 include:_spf.kirinnet.org -all',
      "did:dns:v=1;fp=#{fp_calc};n=QWxpY2U;g=F;iat=#{now};exp=#{now + 3600}",
      "did:dns:pk;kty=ed25519;pk=#{pk_b64}",
      'did:dns:black;fp=RevokedAaaa,RevokedBbbb'
    ]
    did = parse_did_dns_identity(recs)
    raise 'did:dns identity' unless did
    raise 'did:dns version' unless did[:version] == 1
    raise 'did:dns fp' unless did[:fingerprint] == fp_calc
    raise 'did:dns kty' unless did[:key_type] == 'ed25519'
    raise 'did:dns chain' unless fingerprint_chain_ok?(did)
    raise 'did:dns valid' unless did_dns_valid?(did, now)
    raise 'did:dns not revoked' if revoked?(did)

    # Tampered pk -> chain breaks
    wrong_b64 = base64url_encode("\xff" * 32)
    tampered = [recs[0], recs[1], "did:dns:pk;kty=ed25519;pk=#{wrong_b64}"]
    broken = parse_did_dns_identity(tampered)
    raise 'tampered pk breaks chain' if fingerprint_chain_ok?(broken)

    # Missing pk -> nil
    raise 'missing pk -> nil' unless parse_did_dns_identity([recs[1]]).nil?
    # No did:dns -> nil (legacy id= ignored)
    raise 'no did:dns -> nil' unless parse_did_dns_identity(['v=spf1 -all', 'id=foo;key=bar']).nil?
    # Wrong kty -> invalid
    rsa_id = parse_did_dns_identity([recs[1], "did:dns:pk;kty=rsa;pk=#{pk_b64}"])
    raise 'rsa kty rejected' unless rsa_id[:key_type] == 'rsa' && !did_dns_valid?(rsa_id, now)

    puts 'KirinDNS Ruby did:dns self-test: PASSED (fingerprint chain)'
  end
end

File: 02_Libraries/php/kirin_dns.php
<?php
/**
 * KirinDNS Resolution Protocol (ADRP) v2.0 — PHP Client Library
 *
 * Implements ADRP as defined in 01_Standard/spec_v1.md.
 *
 * Architecture:
 *   SRV records for service port discovery (_kirinnet-http._tcp, etc.)
 *   TXT records for identity metadata (id=;key=;nick=;ipfs=)
 *
 * Requires: PHP 8.0+
 */

namespace KirinDNS;

// ---------------------------------------------------------------------------
// Constants (spec Section 2.2)
// ---------------------------------------------------------------------------

const SRV_SERVICES = [
    'http'  => '_kirinnet-http._tcp',
    'https' => '_kirinnet-https._tcp',
    'ws'    => '_kirinnet-ws._tcp',
];

const FALLBACK_PORTS = [
    'http'  => 80,
    'https' => 443,
    'ws'    => 80,
    'wss'   => 443,
];

// ---------------------------------------------------------------------------
// Service Resolution (SRV)
// ---------------------------------------------------------------------------

/**
 * Resolve a single service port via SRV.
 *
 * @return array{target: string, port: int}|null
 */
function resolveService(string $domain, string $service): ?array
{
    $srvName = SRV_SERVICES[$service] ?? null;
    if ($srvName === null) {
        throw new \InvalidArgumentException(
            "Unknown service: $service. Recognized: http, https, ws"
        );
    }

    $fullName = "{$srvName}.{$domain}";
    $records = @dns_get_record($fullName, DNS_SRV);
    if ($records === false || count($records) === 0) {
        return null;
    }

    // RFC 2782: lowest priority, then highest weight
    usort($records, function (array $a, array $b): int {
        if ($a['pri'] !== $b['pri']) return $a['pri'] - $b['pri'];
        return $b['weight'] - $a['weight'];
    });

    $best = $records[0];
    return [
        'target' => rtrim($best['target'], '.'),
        'port'   => $best['port'],
    ];
}

/**
 * Resolve all SRV services for a domain.
 *
 * @return array{http: ?array, https: ?array, ws: ?array}
 */
function resolveAllServices(string $domain): array
{
    return [
        'http'  => resolveService($domain, 'http'),
        'https' => resolveService($domain, 'https'),
        'ws'    => resolveService($domain, 'ws'),
    ];
}

// ---------------------------------------------------------------------------
// Identity Resolution (TXT)
// ---------------------------------------------------------------------------

/**
 * Parse a semicolon-separated key=value TXT string into an identity array.
 *
 * Format: id=<uuid>;key=<hex>;nick=<name>;ipfs=<bool>
 * (spec Section 3.2)
 *
 * @return array{id: string, key: string, nick?: string, ipfs?: bool}|null
 */
function parseIdentityTxt(string $txt): ?array
{
    $txt = trim($txt);
    if ($txt === '' || !str_starts_with($txt, 'id=')) {
        return null;
    }

    $result = [];
    foreach (explode(';', $txt) as $pair) {
        $eq = strpos($pair, '=');
        if ($eq === false) continue;
        $key = trim(substr($pair, 0, $eq));
        $val = trim(substr($pair, $eq + 1));
        $result[$key] = $val;
    }

    // Both id and key are required
    if (!isset($result['id']) || !isset($result['key'])) {
        return null;
    }

    // Parse ipfs boolean
    if (isset($result['ipfs'])) {
        $result['ipfs'] = ($result['ipfs'] === 'true');
    }

    return $result;
}

/**
 * Resolve identity metadata from TXT record.
 *
 * @return array{id: string, key: string, nick?: string, ipfs?: bool}|null
 */
function resolveIdentity(string $domain): ?array
{
    $records = @dns_get_record($domain, DNS_TXT);
    if ($records === false || count($records) === 0) {
        return null;
    }

    foreach ($records as $record) {
        $txt = $record['txt'] ?? ($record['entries'][0] ?? '');
        $identity = parseIdentityTxt($txt);
        if ($identity !== null) {
            return $identity;
        }
    }

    return null;
}

// ---------------------------------------------------------------------------
// did:dns three-record identity model (spec §3.2.1 / did-dns-protocol §2)
// ---------------------------------------------------------------------------
//
// Tamper-evident fingerprint chain: fp == Base64URL(SHA-256(pk)[0:12]).
// Uses PHP's built-in hash() (SHA-256) and a base64url wrapper.

const DID_DNS_PREFIX       = 'did:dns:';
const DID_DNS_DECL_PREFIX  = 'did:dns:v=';
const DID_DNS_PK_PREFIX    = 'did:dns:pk;';
const DID_DNS_BLACK_PREFIX = 'did:dns:black;';
const DID_DNS_KTY_ED25519  = 'ed25519';
const DID_DNS_FINGERPRINT_BYTES = 12;  // -> 16 base64url chars

/** Base64URL encode (RFC 4648 §5, no padding). */
function base64urlEncode(string $binary): string
{
    return rtrim(strtr(base64_encode($binary), '+/', '-_'), '=');
}

/** Base64URL decode (no padding). Returns the raw binary, or null on error. */
function base64urlDecode(string $s): ?string
{
    $pad = strlen($s) % 4;
    if ($pad > 0) {
        $s .= str_repeat('=', 4 - $pad);
    }
    $decoded = base64_decode(strtr($s, '-_', '+/'), true);
    return $decoded === false ? null : $decoded;
}

/**
 * A parsed did:dns identity (declaration + public key, optional blacklist).
 * Returned as an associative array; the tamper-evident fingerprint chain is
 * checked by fingerprintChainOk(), and the caller applies trust policy via
 * isValid() (fail-closed default).
 */
function newDidDnsIdentity(): array
{
    return [
        'version'          => 1,
        'fingerprint'      => '',
        'nickname'         => '',      // Base64URL(UTF-8)
        'gender'           => '',      // M/F/O/X
        'issued_at'        => 0,
        'expires_at'       => 0,
        'key_type'         => DID_DNS_KTY_ED25519,
        'public_key_b64url'=> '',
        'blacklist'        => [],
    ];
}

/** Recompute fp = Base64URL(SHA-256(pk)[0:12]). Empty string on malformed pk. */
function computeFingerprint(array $id): string
{
    $pk = base64urlDecode($id['public_key_b64url'] ?? '');
    if ($pk === null || $pk === '') {
        return '';
    }
    $digest = hash('sha256', $pk, binary: true);
    return base64urlEncode(substr($digest, 0, DID_DNS_FINGERPRINT_BYTES));
}

/** True iff declared fp matches recomputed fp over the pk bytes. */
function fingerprintChainOk(array $id): bool
{
    return $id['fingerprint'] !== '' && $id['fingerprint'] === computeFingerprint($id);
}

function isRevoked(array $id): bool
{
    return in_array($id['fingerprint'], $id['blacklist'], true);
}

function isExpired(array $id, int $now): bool
{
    return $id['expires_at'] !== 0 && $now >= $id['expires_at'];
}

/** Composite policy: v1 + ed25519 + chain + not revoked + not expired. */
function isValid(array $id, int $now): bool
{
    return $id['version'] === 1
        && $id['key_type'] === DID_DNS_KTY_ED25519
        && fingerprintChainOk($id)
        && !isRevoked($id)
        && !isExpired($id, $now);
}

/** Parse a `k=v;k=v` segment (after the did:dns: prefix) into an assoc array. */
function parseDidDnsKv(string $segment): array
{
    $out = [];
    foreach (explode(';', $segment) as $pair) {
        $eq = strpos($pair, '=');
        if ($eq === false) {
            continue;
        }
        $k = trim(substr($pair, 0, $eq));
        $v = trim(substr($pair, $eq + 1));
        $out[$k] = $v;
    }
    return $out;
}

/**
 * Classify TXT records by did:dns: sub-type; assemble an identity.
 *
 * @param array<int,string> $txtRecords
 * @return array|null  null if no did:dns records / declaration+pk missing.
 */
function parseDidDnsIdentity(array $txtRecords): ?array
{
    $declRaw = $pkRaw = $blackRaw = null;
    foreach ($txtRecords as $raw) {
        $s = trim((string) $raw);
        if ($declRaw === null && str_starts_with($s, DID_DNS_DECL_PREFIX)) {
            $declRaw = $s;
        } elseif ($pkRaw === null && str_starts_with($s, DID_DNS_PK_PREFIX)) {
            $pkRaw = $s;
        } elseif ($blackRaw === null && str_starts_with($s, DID_DNS_BLACK_PREFIX)) {
            $blackRaw = $s;
        }
    }
    if ($declRaw === null || $pkRaw === null) {
        return null;
    }

    $decl = parseDidDnsKv(substr($declRaw, strlen(DID_DNS_PREFIX)));
    $pk   = parseDidDnsKv(substr($pkRaw, strlen(DID_DNS_PREFIX)));

    $id = newDidDnsIdentity();
    $id['version']           = isset($decl['v']) ? (int)$decl['v'] : 1;
    $id['fingerprint']       = $decl['fp'] ?? '';
    $id['nickname']          = $decl['n'] ?? '';
    $id['gender']            = $decl['g'] ?? '';
    $id['issued_at']         = isset($decl['iat']) ? (int)$decl['iat'] : 0;
    $id['expires_at']        = isset($decl['exp']) ? (int)$decl['exp'] : 0;
    $id['key_type']          = $pk['kty'] ?? DID_DNS_KTY_ED25519;
    $id['public_key_b64url'] = $pk['pk'] ?? '';

    if ($blackRaw !== null) {
        $bkv = parseDidDnsKv(substr($blackRaw, strlen(DID_DNS_PREFIX)));
        $fpField = $bkv['fp'] ?? '';
        $id['blacklist'] = array_values(array_filter(explode(',', $fpField), fn ($f) => $f !== ''));
    }
    return $id;
}

// ---------------------------------------------------------------------------
// Legacy Compatibility Wrapper
// ---------------------------------------------------------------------------

/**
 * Full resolution: SRV + TXT + identity (legacy wrapper).
 *
 * New code should use resolveService() and resolveIdentity() directly.
 */
function resolve_kirin_dns(string $domain): array
{
    $ws = resolveService($domain, 'ws');

    return [
        'domain'   => $domain,
        'ws'       => $ws ?? ['target' => $domain, 'port' => FALLBACK_PORTS['ws']],
        'http'     => resolveService($domain, 'http'),
        'https'    => resolveService($domain, 'https'),
        'identity' => resolveIdentity($domain),
    ];
}

// ---------------------------------------------------------------------------
// Self-test (run: php kirin_dns.php)
// ---------------------------------------------------------------------------
if (basename(__FILE__) === basename($_SERVER['SCRIPT_FILENAME'] ?? '')) {
    // SRV — nonexistent domain returns null
    $ws = resolveService('nonexistent.invalid', 'ws');
    assert($ws === null, 'no SRV for nonexistent domain');

    // TXT identity — nonexistent domain returns null
    $id = resolveIdentity('nonexistent.invalid');
    assert($id === null, 'no TXT identity for nonexistent domain');

    // Identity parser
    $parsed = parseIdentityTxt(
        'id=550e8400-e29b-41d4-a716-446655440000;key=04abc;nick=Alice;ipfs=false'
    );
    assert($parsed['id']   === '550e8400-e29b-41d4-a716-446655440000', 'id parsed');
    assert($parsed['key']  === '04abc', 'key parsed');
    assert($parsed['nick'] === 'Alice', 'nick parsed');
    assert($parsed['ipfs'] === false, 'ipfs parsed as bool');

    $minimal = parseIdentityTxt('id=test-id;key=0x00');
    assert($minimal['id']  === 'test-id', 'minimal id');
    assert($minimal['key'] === '0x00', 'minimal key');
    assert(!isset($minimal['nick']), 'no nick');

    // Invalid TXT
    assert(parseIdentityTxt('v=spf1 include:_spf.example.com') === null, 'spf skipped');
    assert(parseIdentityTxt('') === null, 'empty string');
    assert(parseIdentityTxt('not an identity') === null, 'not identity');

    // Legacy wrapper
    $full = resolve_kirin_dns('nonexistent.invalid');
    assert($full['ws']['port'] === 80, 'legacy ws fallback');
    assert($full['http'] === null, 'legacy http null');
    assert($full['identity'] === null, 'legacy identity null');

    echo "KirinDNS PHP self-test: PASSED\n";

    // ---- did:dns three-record identity model (C-1 baseline) ----
    // Deterministic 32-byte key = bytes 0..31 (matches the golden vector).
    $pkBytes = pack('C*', ...range(0, 31));
    $pkB64 = base64urlEncode($pkBytes);
    $dg = hash('sha256', $pkBytes, binary: true);
    $fpCalc = base64urlEncode(substr($dg, 0, DID_DNS_FINGERPRINT_BYTES));
    $now = 1700000000;

    $recs = [
        'v=spf1 include:_spf.kirinnet.org -all',
        "did:dns:v=1;fp={$fpCalc};n=QWxpY2U;g=F;iat={$now};exp=" . ($now + 3600),
        "did:dns:pk;kty=ed25519;pk={$pkB64}",
        'did:dns:black;fp=RevokedAaaa,RevokedBbbb',
    ];
    $did = parseDidDnsIdentity($recs);
    assert($did !== null, 'did:dns identity parsed');
    assert($did['version'] === 1, 'did:dns version');
    assert($did['fingerprint'] === $fpCalc, 'did:dns fp');
    assert($did['key_type'] === 'ed25519', 'did:dns kty');
    assert(fingerprintChainOk($did), 'did:dns fingerprint chain');
    assert(isValid($did, $now), 'did:dns valid');
    assert(!isRevoked($did), 'did:dns not revoked');

    // Tampered pk -> chain breaks
    $wrongB64 = base64urlEncode(str_repeat("\xff", 32));
    $tampered = [$recs[0], $recs[1], "did:dns:pk;kty=ed25519;pk={$wrongB64}"];
    $broken = parseDidDnsIdentity($tampered);
    assert(!fingerprintChainOk($broken), 'tampered pk breaks chain');

    // Missing pk -> null
    assert(parseDidDnsIdentity([$recs[1]]) === null, 'missing pk -> null');
    // No did:dns -> null (legacy id= ignored)
    assert(parseDidDnsIdentity(['v=spf1 -all', 'id=foo;key=bar']) === null, 'no did:dns -> null');
    // Wrong kty -> invalid
    $rsaId = parseDidDnsIdentity([$recs[1], "did:dns:pk;kty=rsa;pk={$pkB64}"]);
    assert($rsaId['key_type'] === 'rsa' && !isValid($rsaId, $now), 'rsa kty rejected');

    echo "KirinDNS PHP did:dns self-test: PASSED (fingerprint chain)\n";
}
